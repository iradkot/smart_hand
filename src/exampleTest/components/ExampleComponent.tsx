import React, { useState } from 'react';
import styled from 'styled-components'; // Import styled-components

interface ExampleComponentProps {
  initialCount?: number;
  isDisabled?: boolean;
}

// Styled component example for the container
const Container = styled.div`
  padding: 20px;
  background-color: #f4f4f4;
  border-radius: 8px;
`;

const StyledButton = styled.button`
  margin: 5px;
  padding: 10px 20px;
  font-size: 16px;
  border: none;
  border-radius: 5px;
  background-color: #007bff;
  color: white;
  cursor: pointer;

  &:disabled {
    background-color: grey;
  }
`;

export const ExampleComponent: React.FC<ExampleComponentProps> = ({
                                                                    initialCount = 0,
                                                                    isDisabled = false,
                                                                  }) => {
  const [count, setCount] = useState(initialCount);

  return (
    <Container>
      {/* Accessible name for better testing */}
      <label htmlFor="count">Counter</label>
      <p id="count" data-testid="count-display">Count: {count}</p>

      {/* Increment and decrement buttons */}
      <StyledButton
        onClick={() => setCount(count + 1)}
        data-testid="increment-button"
        disabled={isDisabled}
        type="button" // Set button type explicitly
      >
        Increment
      </StyledButton>
      <StyledButton
        onClick={() => setCount(count - 1)}
        data-testid="decrement-button"
        disabled={isDisabled}
        type="button" // Set button type explicitly
      >
        Decrement
      </StyledButton>
    </Container>
  );
};
