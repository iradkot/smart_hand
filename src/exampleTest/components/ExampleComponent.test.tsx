// examples/components/ExampleComponent.test.tsx
import { render, fireEvent, screen } from '@testing-library/react';
import { ExampleComponent } from './ExampleComponent';

// Helper function to set up component for tests
const setup = (props = {}) => {
  render(<ExampleComponent {...props} />);
  const incrementButton = screen.getByTestId('increment-button');
  const decrementButton = screen.getByTestId('decrement-button');
  const countDisplay = screen.getByTestId('count-display');
  return { incrementButton, decrementButton, countDisplay };
};

describe('ExampleComponent', () => {

  describe('Initial Rendering', () => {
    test('renders with default initial count', () => {
      const { countDisplay, incrementButton, decrementButton } = setup();
      expect(countDisplay).toHaveTextContent('Count: 0'); // Default count
      expect(incrementButton).toBeEnabled();
      expect(decrementButton).toBeEnabled();
    });

    test.each`
      initialCount | expectedText
      ${0}         | ${'Count: 0'}
      ${5}         | ${'Count: 5'}
    `('renders correctly with initialCount=$initialCount', ({ initialCount, expectedText }) => {
      const { countDisplay } = setup({ initialCount });
      expect(countDisplay).toHaveTextContent(expectedText);
    });
  });

  describe('Button Interactions', () => {
    test('increments count when increment button is clicked', () => {
      const { incrementButton, countDisplay } = setup();
      fireEvent.click(incrementButton);
      expect(countDisplay).toHaveTextContent('Count: 1');
    });

    test('decrements count when decrement button is clicked', () => {
      const { decrementButton, countDisplay } = setup({ initialCount: 1 });
      fireEvent.click(decrementButton);
      expect(countDisplay).toHaveTextContent('Count: 0');
    });
  });

  describe('Disabled Buttons', () => {
    test('disables increment and decrement buttons when isDisabled prop is true', () => {
      const { incrementButton, decrementButton } = setup({ isDisabled: true });
      expect(incrementButton).toBeDisabled();
      expect(decrementButton).toBeDisabled();
    });
  });

  describe('Accessibility and ARIA', () => {
    test('increment button has accessible name', () => {
      const { incrementButton } = setup();
      expect(incrementButton).toHaveAccessibleName('Increment');
    });

    test('decrement button has accessible name', () => {
      const { decrementButton } = setup();
      expect(decrementButton).toHaveAccessibleName('Decrement');
    });
  });

  describe('Visual and Focus Tests', () => {
    test('count display is visible', () => {
      const { countDisplay } = setup();
      expect(countDisplay).toBeVisible();
    });

    test('increment button can receive focus', () => {
      const { incrementButton } = setup();
      incrementButton.focus();
      expect(incrementButton).toHaveFocus();
    });

    test('matches snapshot for visual consistency', () => {
      const { container } = render(<ExampleComponent />);
      expect(container).toMatchSnapshot();
    });
  });

  describe('Edge Cases', () => {
    test('handles negative initial count', () => {
      const { countDisplay } = setup({ initialCount: -5 });
      expect(countDisplay).toHaveTextContent('Count: -5');
    });

    test('renders with missing optional props', () => {
      const { countDisplay } = setup();
      expect(countDisplay).toHaveTextContent('Count: 0');
    });
  });

});
