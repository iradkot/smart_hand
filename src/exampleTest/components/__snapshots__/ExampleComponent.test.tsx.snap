// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExampleComponent Visual and Focus Tests matches snapshot for visual consistency 1`] = `
<div>
  <div
    class="sc-braxZu cSCVkU"
  >
    <label
      for="count"
    >
      Counter
    </label>
    <p
      data-testid="count-display"
      id="count"
    >
      Count: 
      0
    </p>
    <button
      class="sc-gJhJTp NAVpP"
      data-testid="increment-button"
      type="button"
    >
      Increment
    </button>
    <button
      class="sc-gJhJTp NAVpP"
      data-testid="decrement-button"
      type="button"
    >
      Decrement
    </button>
  </div>
</div>
`;
